spec/const3.dd

```diff
@@ -262,14 +262,25 @@ char[] p = "world"; // error, cannot implicitly convert immutable
 
         $(P
         The second way is to cast data to immutable.
-        When doing so, it is up to the programmer to ensure that no
-        other mutable references to the same data exist.
+        When doing so, it is up to the programmer to ensure that any mutable
+        references to the same data are not used to modify the data after the
+        cast.
         )
 
 ---
-char[] s = ...;
-immutable(char)[] p = cast(immutable)s;     // undefined behavior
-immutable(char)[] p = cast(immutable)s.dup; // ok, unique reference
+char[] s = ['a'];
+s[0] = 'b'; // ok
+immutable(char)[] p = cast(immutable)s; // ok, if data is not mutated
```

Comment:
I'd still prefer to make this `accessed through s` – is there any tangible benefit to relaxing this restriction?

Reply:
The reason for my preference (although it has admittedly been a while since I thought about this) is conceptual simplicity: If casting invalidates all old references, then any object is always in one of two states that are straightforwardly defined: mutable or immutable.

Immutable objects can be assumed not to be modified, mutable objects can be assumed not to be shared (in theory, anyway). If there can be both mutable and immutable references at the same time, there is a third state where neither guarantee applies.

(Whether it was wise to make immutable implicitly sharable is a question for another day, of course.)

Reply:
> I'd still prefer to make this `accessed through s` – is there any tangible benefit to relaxing this restriction?

The goal is to fix issue [19621](https://issues.dlang.org/show_bug.cgi?id=19621). That can be done by either relaxing the rules for mutable → immutable (as done here), or by tightening the rules for immutable → mutable.

If I keep the rules tight here, we have to tighten them in the other direction as well, because both directions are equally dangerous. I.e., if I write "accessed" here, then the following kind of access should also be illegal, but it's currently [allowed by the spec](https://dlang.org/spec/const3.html#removing_with_cast):

```d
int main()
{
    immutable int* i = new int(42);
    return f(i, cast(int*) i); /* immutable → mutable */
}
int f(immutable int* i, int* m)
{
    return *i + *m; /* the spec currently says (at least implies) that this is ok */
}
```

I guess we could make that illegal, but it would be a silent, subtle breaking change, as far as I see. The benefit of relaxing the rules here is to avoid that breaking change. I don't know if it outweighs the benefits of forbidding the aliasing.

Reply:
I don't understand your f() example. What part of the code is @trusted?

Reply:
> I don't understand your f() example. What part of the code is @trusted?

There is no `@trusted` code. This discussion is about where exactly undefined behavior is triggered. That applies to `@system` code, too.

If a compiler looks at `f` in isolation, can it assume that the parameters `i` and `m` won't both be used to access the same data? @dnadlinger would like the answer to be "yes". My example is supposed to show that the answer must be "no" with the current spec, regardless of the changes in this PR.